Tidy InvitationsController

floehopper opened the pull request:
Trello: https://trello.com/c/umBDQZUc

This is in preparation for splitting the page handled by the `InvitationsController#new` & `#create` actions into two pages with the second handling the setting of permissions. The `InvitationsController` is made more complicated than a typical controller for a few reasons, e.g.:

1.  The audience for the `#new`, `#create` & `#destroy` actions is admins & superadmins inviting new users; whereas the audience for the `#edit` & `#update` actions is new users accepting their invitations and setting their password.
2. It inherits from [`Devise::InvitationsController`][1] provided by `devise_invitable` but significantly customises its behaviour and its template(s) in `app/views/devise/invitations/`.

There's definitely a question in my mind about how much benefit we're getting from inheriting from `Devise::InvitationsController`, particularly given how much we are customising it and how easy it is for the code to get out-of-date with the gem as it gets updated with Dependabot. However, I decided that this is out-of-scope for this PR.

Currently there's an optional 2nd page for when an admin can choose whether or not the invited user must use 2SV which further complicates the controller. I'm hoping that I can combine this page into the 1st page in a subsequent PR.

While this PR has a lot of commits, most of them are relatively small and only make changes to two files:
* `app/controllers/invitations_controller.rb`
* `test/controllers/invitations_controller_test.rb`

The main changes in this PR are:
* Improve test coverage in `InvitationsControllerTest` and try to make `InvitationsControllerTest` easier to follow.
* Simplify `InvitationsController` and bring it more into line with the current version of `Devise::InvitationsController`
    * In particular, make use of [the built-in support for Strong Parameters][2] which I suspect may not have been available when some of the code was written.
    * As part of this, the controller no longer uses `UserParameterSanitiser`. I'm not convinced that the latter is a useful abstraction, because it makes it much harder to see which parameters are permitted for a given controller/action and (in combination with `Roles::***.permitted_user_params` methods it seems to contain logic that probably belongs in Pundit policies.
* Introduce a separate `app/views/devise/invitations/new.html.erb` template rather than re-using `app/views/users/_form_fields.html.erb` - the latter was unnecessarily complicated and made it hard to see the wood for the trees.  Admittedly this comes at the cost of a little duplication, but I think the benefits outweigh the costs.

[1]: https://github.com/scambra/devise_invitable/blob/v2.0.8/app/controllers/devise/invitations_controller.rb
[2]: https://github.com/scambra/devise_invitable/blob/v2.0.8/README.rdoc#label-Strong+Parameters

chrisroos reviewed (approved):
This all looks good to me, @floehopper!

I've left a couple of comments in the code but I'll leave you to decide whether to address them.

The two commits I most struggled to understand were `Use built-in strong params for InvitationsController#update` and `Use built-in strong params for InvitationsController#create` but I trust you and the tests that the behaviour is still as expected :-)

I also noticed a few improvements that you could make to commit notes, but again I'll leave it to decide whether to address them:

## Make InvitationsController more like its superclass

> The superclass, Devise::InvitationsController, prepends the
> `authenticate_inviter!` before action [1] for the `new` & `create`
> actions. This before action calls `authenticate_user!` anyway [2],
> albeit with the `force` option set to `true`.

The [2] links to https://github.com/scambra/devise_invitable/blob/db2f9a1c0e87addb196d1b82920370f4230c1157/app/controllers/devise/invitations_controller.rb#L4 which doesn't seem to make sense in this context.

> So, all in all, I think it's clearer and more consistent with the
> superclass implementation to add `require_no_authentication!` as a
> before action for `new`, `create` AND `resend` actions.

I read this assuming that I'd see a call to `require_no_authentication!` in this commit but there isn't one. Does it instead mean that we're relying on the call to `require_no_authentication!` in the superclass?

---

## Rename "user" to "inviter" in InvitationsControllerTest

> named "user". I've changed the relevant instances of "user" to "invitee"

I think "invitee" should be "inviter".

---

## Extract more contexts in InvitationsControllerTest

> to require 2S when the invitee's organisation does not require 2SV.

Missing "V" from "2SV".

---

## Make InvitationsControllerTest test names more explanatory

> they will be assined to an organisation for which 2SV is not required,

Typo in "assigned".

---

## Use built-in strong params for InvitationsController#create

Lack of line breaks makes it hard to read.

floehopper commented:
@chrisroos Many thanks for all your feedback. See my responses inline below:

> This all looks good to me, @floehopper!
> 
> I've left a couple of comments in the code but I'll leave you to decide whether to address them.
> 
> The two commits I most struggled to understand were `Use built-in strong params for InvitationsController#update` and `Use built-in strong params for InvitationsController#create` but I trust you and the tests that the behaviour is still as expected :-)
> 
> I also noticed a few improvements that you could make to commit notes, but again I'll leave it to decide whether to address them:
> ## Make InvitationsController more like its superclass
> 
> > The superclass, Devise::InvitationsController, prepends the
> > `authenticate_inviter!` before action [1] for the `new` & `create`
> > actions. This before action calls `authenticate_user!` anyway [2],
> > albeit with the `force` option set to `true`.
> 
> The [2] links to https://github.com/scambra/devise_invitable/blob/db2f9a1c0e87addb196d1b82920370f4230c1157/app/controllers/devise/invitations_controller.rb#L4 which doesn't seem to make sense in this context.

Oh, yes - well spotted - the links were in quite a muddle. I've hopefully fixed them now!

> > So, all in all, I think it's clearer and more consistent with the
> > superclass implementation to add `require_no_authentication!` as a
> > before action for `new`, `create` AND `resend` actions.
> 
> I read this assuming that I'd see a call to `require_no_authentication!` in this commit but there isn't one. Does it instead mean that we're relying on the call to `require_no_authentication!` in the superclass?

No, it was a typo. `require_no_authentication!` should've been `authenticate_user!` in this paragraph. No wonder you were confused! I've fixed the typo and expanded the paragraph a bit to hopefully make things clearer.

> ## Rename "user" to "inviter" in InvitationsControllerTest
> 
> > named "user". I've changed the relevant instances of "user" to "invitee"
> 
> I think "invitee" should be "inviter".

Yes, you're correct - I've fixed the typo.

> ## Extract more contexts in InvitationsControllerTest
> 
> > to require 2S when the invitee's organisation does not require 2SV.
> 
> Missing "V" from "2SV".

Yes, I've fixed the typo.

> ## Make InvitationsControllerTest test names more explanatory
> 
> > they will be assined to an organisation for which 2SV is not required,
> 
> Typo in "assigned".

Typo fixed.

> ## Use built-in strong params for InvitationsController#create
> 
> Lack of line breaks makes it hard to read.

Ah, yes - sorry - now fixed!

I'll address your inline comments before I force-push the changes.

floehopper commented on test/controllers/invitations_controller_test.rb:
```diff
@@ -146,7 +146,9 @@ class InvitationsControllerTest < ActionController::TestCase
 
-      should "re-render form and not save user if there are validation errors" do
+      should "re-render form and not save invitee if there are validation errors" do
+        organisation = create(:organisation)
```
Nope - now fixed! 👍

floehopper commented on test/policies/user_policy_test.rb:
```diff
@@ -12,8 +12,8 @@ class UserPolicyTest < ActiveSupport::TestCase
-  primary_management_actions = %i[new assign_organisations]
-  user_management_actions = %i[edit create update unlock suspension cancel_email_change resend_email_change event_logs reset_2sv mandate_2sv]
+  primary_management_actions = %i[new create assign_organisations]
+  user_management_actions = %i[edit update unlock suspension cancel_email_change resend_email_change event_logs reset_2sv mandate_2sv]
```
Done 👍

floehopper commented on test/controllers/invitations_controller_test.rb:
```diff
@@ -78,93 +78,91 @@ class InvitationsControllerTest < ActionController::TestCase
       end
 
       should "re-render form and not save invitee if there are validation errors" do
-        organisation = create(:organisation)
```
Yep - now resolved 👍

floehopper commented:
Force-pushing the changes I've made in the light of @chrisroos' feedback in preparation for merging.
